File: WebApplication/ClientApp/src/app/_orchestration/getRaceSeriesDashboard/raceSeriesDashboardDto.ts
import { LocationInfoWithRank } from "../locationInfoWithRank";
import { CourseInformationEntry } from "./courseInformationEntry";
import { DisplayNameWithIdDto } from "../displayNameWithIdDto";
import { IRaceSeriesType } from "../../_common/IRaceSeriesType";

export interface RaceSeriesDashboardDto extends IRaceSeriesType {
  name: string
  description: string
  kickOffDate: string
  races: PastRaceDto[]
  courses: RaceSeriesDashboardCourseDto[]
  locationInfoWithRank: LocationInfoWithRank
  upcomingRaceId: number
  firstCourseId: number
}

export interface RaceSeriesDashboardCourseDto extends DisplayNameWithIdDto {
  descriptionEntries: CourseInformationEntry[]
  promotionalEntries: CourseInformationEntry[]
  howToPrepareEntries: CourseInformationEntry[]
  participants: RaceSeriesDashboardParticipantDto[]
}

export interface RaceSeriesDashboardParticipantDto {
  athleteId: number
  athleteCourseId: number
  fullName: string
  firstName: string
  bib: string
  state: string
  city: string
  raceAge: number
  genderAbbreviated: string
  courseGoalDescription: string
  personalGoalDescription: string
  trainingList: string[]
}

export interface PastRaceDto extends DisplayNameWithIdDto {
  kickOffDate: string
}

File: Api/Orchestration/GetRaceSeriesDashboard/RaceSeriesDashboardDto.cs
using Api.DataModels;
using Api.DataModels.Enums;

namespace Api.Orchestration.GetRaceSeriesDashboard;

public static class RaceSeriesDashboardDtoMapper
{
    public static RaceSeriesDashboardDto GetRaceSeriesDashboardDto(RaceSeries raceSeries, List<PastRaceDto> races, List<RaceSeriesDashboardCourseDto> courses)
    {
        var upcomingRace = races.First();
        var firstCourse = courses.First();

        return new
        (
            raceSeries.Name,
            raceSeries.Description,
            upcomingRace.KickOffDate,
            races,
            courses,
            raceSeries.RaceSeriesType,
            new LocationInfoWithRank(raceSeries),
            upcomingRace.Id,
            firstCourse.Id
        );
    }
}

public record RaceSeriesDashboardDto
(
    string Name,
    string Description,
    string KickOffDate,
    List<PastRaceDto> Races,
    List<RaceSeriesDashboardCourseDto> Courses,
    RaceSeriesType RaceSeriesType,
    LocationInfoWithRank LocationInfoWithRank,
    int UpcomingRaceId,
    int FirstCourseId
);

